src/promnesia/sources/viber_desktop.py

```diff
@@ -0,0 +1,164 @@
+            UNION
+            SELECT ChatId, chatname, PGTags FROM G2
+        )
+        SELECT
+            M.EventId           AS mid,
+            E.TimeStamp         AS time,
+            G.chatname          AS chatname,
+            coalesce(
+                S.Name,
+                S.ClientName,
+                '(' || S.Number || ')'          /* contacts have one xor the other, but failsafe */
+            )                   AS sender,
+            coalesce(M.Subject, M.Body)         /* didn't see any msg with both */    
+                                AS text,
+            M.info              AS infojson,    /* to harvested titles from embedded urls */
+            G.PGTags            AS tags
+        FROM messages AS M
+        LEFT JOIN Events AS E
+            ON M.EventId = E.EventId
+        LEFT JOIN Contact AS S
+            ON E.ContactId = S.ContactId
+        LEFT JOIN Groups AS G
+            ON E.ChatId = G.ChatId
+        WHERE
+            text LIKE '%http%'
+        ORDER BY time;
+        """
+    )
+
+    def _parse_json_title(js) -> str:
+        if js and js.strip():
+            js = json.loads(js)
+            if isinstance(js, dict):
+                return js.get("Title")
+
+    def _handle_row(row) -> Results:
+        text = row["text"]
+        assert (
+            text
+        ), f"sql-query should have eliminated messages not containing 'http': {text}"
+        urls = extract_urls(text)
+        if not urls:
+            return
+        dt = from_epoch(row["time"] // 1000)  # timestamps are stored x100 this db
+        mid: str = unwrap(row["mid"])
+        # TODO perhaps we could be defensive with null sender/chat etc and still emit the Visit
+        sender: str = unwrap(row["sender"])
+        chatname: str = unwrap(row["chatname"])
+        sender: str = unwrap(row["sender"])
+        tags: str = unwrap(row["tags"])
+        infojson: str = unwrap(row["infojson"])
+
+        if tags and tags.strip():
+            tags = "".join(f"#{t}" for t in tags.split())
+            text = f"{text}\n\n{tags}"
+
+        url_title = _parse_json_title(infojson)
+        if url_title:
+            text = f"title: {url_title}\n\n{text}"
+
+        for u in urls:
+            # https://www.reddit.com/r/Telegram/comments/6ufwi3/link_to_a_specific_message_in_a_channel_possible/
+            # hmm, only seems to work on mobile app, but better than nothing...
+            yield Visit(
+                url=unquote(u),
+                dt=dt,
+                context=text,
+                locator=Loc.make(
+                    title=f"chat({mid}) from {sender}@{chatname}",
+                    href=f"sqlite://{database}?immutable=1#!Messages.EventId={mid}",
+                ),
+            )
+
+    # TODO yield error if chatname or chat or smth else is null?
+    for row in db.query(query_str):
+        try:
+            yield from _handle_row(row)
+        except Exception as ex:
+            logger.warning(
```

Comment:
Replying to your comment [here](https://github.com/karlicoss/promnesia/commit/4363b5797b30819b89a4e4908ea27b009b598112#commitcomment-47433572) 

Basically the options people usually use are:

1. throw immediately (as you suggested in the comment)
  This is problematic because then on any unexpected situation (like 1/10000 messages parsed incorrectly), the program will break. Considering we have to deal with closed protocols, formats change without warning, and it's annoying when your software breaks until you fix it, so we need to be more flexible and defensive here. In addition, imagine if one data source breaks the whole program -- sometimes it's justified (e.g. if you were working with important medical data or something like that), but not in case of this tool.

2. log and continue (as you're doing here)
   It works nice in terms of not breaking immediately. However then there is another danger -- imagine that for example Viber changes format for 'time' column to store the isoformatted string (instead of epoch), for whatever reason. Then promnesia indexer will carry on working (because of exception catching) -- it's good. However, you won't notice the error unless you regularly reviewing the logs, or maybe when you know for sure that Promnesia should fire but it doesn't.

So in Promnesia (and related projects that process data, like HPI) I'm doing something intermediate.

- catch the exceptions (where it makes sense, e.g. parsing)
- possibly log it if necessary
- but then `yield` it to the parent function, as a normal value (there is even an alias for [`Result`](https://github.com/karlicoss/promnesia/blob/9b7520feea9872e29520d466aa302b76ea068c52/src/promnesia/common.py#L128-L130) )
 
This allows the top level handler to exit with code 1 (so you get a cron email/systemd job failure etc and notice the error), while still doing what's the program supposed to do (process as much as it can and insert in the database).

As more examples why it's useful in Promnesia, errors are also inserted in the database:
- this makes it easier to test in some cases (instead of intercepting logs, just check the database)
- later it will allow to overview the errors form the browser which would be much more convenient for most users than messing with systemd logs or whatnote

I write more about it [here](https://beepb00p.xyz/mypy-error-handling.html), you might enjoy it!